lambdas/data-migration1/src/providers.ts

```diff
@@ -0,0 +1,157 @@
+import Knex from 'knex';
+import isNil from 'lodash/isNil';
+
+import DynamoDbSearchQueue from '@cumulus/aws-client/DynamoDbSearchQueue';
+import * as KMS from '@cumulus/aws-client/KMS';
+import { envUtils, keyPairProvider } from '@cumulus/common';
+import Logger from '@cumulus/logger';
+
+import { RecordAlreadyMigrated } from './errors';
+import { MigrationSummary } from './types';
+
+const Manager = require('@cumulus/api/models/base');
+const schemas = require('@cumulus/api/models/schemas');
+
+const logger = new Logger({ sender: '@cumulus/data-migration/providers' });
+
+export interface RDSProviderRecord {
+  name: string
+  protocol: string
+  host: string
+  port?: number
+  username?: string
+  password?: string
+  encrypted?: boolean
+  globalConnectionLimit?: number
+  privateKey?: string
+  cmKeyId?: string
+  certificateUri?: string
+  created_at: Date
+  updated_at?: Date
+}
+
+const encryptProviderCredential = async (
+  value: string,
+  encrypted?: boolean
+) => {
+  if (isNil(value)) return undefined;
+
+  const providerKmsKeyId = envUtils.getRequiredEnvVar('provider_kms_key_id');
+
+  if (encrypted) {
+    return keyPairProvider.S3KeyPairProvider
+      .decrypt(value)
+      .then(
+        (decryptedValue) => KMS.encrypt(providerKmsKeyId, decryptedValue),
+        async (error) => {
```

Comment:
I'd rather see `try / catch` here rather than the less-common `then(resolve, reject)` syntax.

Reply:
This is a bit complex. If the decryption fails and the KMS decryption succeeds, then we want to return the initial passed in value from this function, but otherwise we want to encrypt the decrypted value. This felt cleaner than other options that I tried, but I can revisit.

Reply:
I've tried to refactor this a couple times now and I'm not seeing a great way without the `then(resolve,reject)` syntax. The problem is that we want to attempt verifying that it is a KMS encrypted value **only if** the S3 decryption failed, not if the `resolve` function itself throws an error. And we only need to encrypt the value if the S3 decryption succeeded. So far, I'm not finding an elegant way to achieve this in try/catch syntax. This gets close, but the catch would be triggered for any error on KMS.encrypt as well, which is not what we want.

```javascript
    let decryptedValue;
    try {
      decryptedValue = await decryptProviderCredential(value);
      return KMS.encrypt(providerKmsKeyId, decryptedValue);
    } catch (error) {
      // If we already have a KMS encrypted value, return it.
      if ((await KMS.decryptBase64String(value))) {
        return value;
      }
      throw error;
    }
```

Reply:
OK, got this: 159c7e27f6386873b8b6c14dd34ef4adddf1145a

I'm not sure that I really like it better though

Reply:
Here is an option: https://github.com/nasa/cumulus/pull/1882/files